### DISCUSSION TITLE
Conditional command structure

### DISCUSSION THREAD

**Discussion by rjbell4** [mjansen4857/pathplanner#663] [General]
What advice would you give to making an auto conditional?  We've had a lot of luck with simply defining paths, and connection them together with commands and other paths in autos.  But we would now like to explore some conditional behavior.  For example, if going to get a note, and failing (because it's been removed), getting a different note instead, before returning to a shooting position, etc.  I'm sure it's possible, but I'd like not just to do it, but to do it _well_ (especially as I'm a mentor, so I'd like to give the students good advice).

**Comment by gcschmit** [comment]
The approach that our team settled on this year is to create an auto for each possible branch in the decision tree. (We place all of these related autos in a folder to keep them organized.) We then introduce the conditional logic in the command that we build in the Robot Container class. For example, here is the code for our 4-note source-side auto:

```
Command fourNoteSourceSide =
        Commands.sequence(
            Commands.runOnce(() -> shooter.setShootingPosition(ShootingPosition.SOURCE_SIDE_AUTO)),
            new PathPlannerAuto("Collect 2nd"),
            Commands.either(
                new PathPlannerAuto("Score 2nd Collect 3rd"),
                new PathPlannerAuto("Missed 2nd Collect 3rd"),
                intake::hasNote),
            Commands.runOnce(
                () -> shooter.setShootingPosition(ShootingPosition.SOURCE_SIDE_UNDER_STAGE_AUTO)),
            Commands.either(
                new PathPlannerAuto("Score 3rd Collect 4th"),
                new PathPlannerAuto("Missed 3rd Collect 4th"),
                intake::hasNote),
            Commands.either(
                new PathPlannerAuto("Score 4th Center"), Commands.none(), intake::hasNote));
    autoChooser.addOption("4 Note Source Side", fourNoteSourceSide);
```

**Reply by rjbell4** [reply]
Thanks!  So far, all series of actions have been fully defined in Autos, and we simply use `AutoBuilder.buildAutoChooser()` to load and choose them.  It looks like we will want to consider expanding that list with some code-defined sequences that can vary instead.  Thank you for such an applicable example!
